### PR TITLE
Don't log a stack trace for AddressInUseException

### PR DESCRIPTION
**Pull request by BerserkerDotNet**
In case of AddressInUseException do not log the exception
**AddressInUseException:**
image
image

**Other exceptions:**
image

Addresses #29801

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=30154) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=30154) before we can accept your contribution.<br/><br/>:x: BerserkerDotNet [sign now](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=30154)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/aspnetcore?pullRequest=30154) it.</sub>

File: src/Servers/Kestrel/Core/src/Internal/KestrelServerImpl.cs [hunk 1]
```diff
@@ -207,7 +208,16 @@ async Task OnBind(ListenOptions options)
             }
             catch (Exception ex)
             {
-                Trace.LogCritical(0, ex, "Unable to start Kestrel.");
+                // Do not log stack trace for known errors #29801
```
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Internal/KestrelServerImpl.cs:211]: Code and issues have a bad habit of moving around.
```suggestion
                // Do not log stack trace for known errors https://github.com/dotnet/aspnetcore/issues/29801
```

File: src/Servers/Kestrel/test/BindTests/AddressRegistrationTests.cs [hunk 2]
```diff
@@ -578,7 +584,7 @@ public async Task ThrowsWhenBindingToIPv4AddressInUse()
         [IPv6SupportedCondition]
         public async Task ThrowsWhenBindingToIPv6AddressInUse()
         {
-            IgnoredCriticalLogExceptions.Add(typeof(IOException));
+            ThrowOnCriticalErrors = false;
```
Comment (Tratcher) [inline src/Servers/Kestrel/test/BindTests/AddressRegistrationTests.cs:587]: Why change this approach?

**Comment by davidfowl** [comment]
Wait why are we taking this change? Isn't this in the wrong layer? Removing information from loggers is the wrong place to do this sort of filtering.

File: src/Servers/Kestrel/test/BindTests/AddressRegistrationTests.cs (hunk 2 above)
Comment (BerserkerDotNet) [reply src/Servers/Kestrel/test/BindTests/AddressRegistrationTests.cs:587]: There is no exception logged for this even anymore that is why the `IgnoredCriticalLogExceptions` will not work. I'm checking the message content instead.

**Comment by BerserkerDotNet** [comment]
> Wait why are we taking this change? Isn't this in the wrong layer? Removing information from loggers is the wrong place to do this sort of filtering.

Won't argue, it looks weird at best. :) But my understanding is #29801 asks to remove this information from the log without suppressing the exception. Maybe I misunderstood the ask.

**Comment by BrennanConroy** [comment]
Triage: We think we shouldn't log and just let the exception flow up to the user.

**Comment by BerserkerDotNet** [comment]
> Triage: We think we shouldn't log and just let the exception flow up to the user.

Ok, I'll make that change.

**Comment by BerserkerDotNet** [comment]
@BrennanConroy, done. Updated PR description as well.

File: src/Servers/Kestrel/Core/src/Internal/KestrelServerImpl.cs [hunk 3]
```diff
@@ -207,7 +208,12 @@ async Task OnBind(ListenOptions options)
             }
             catch (Exception ex)
             {
-                Trace.LogCritical(0, ex, "Unable to start Kestrel.");
+                // Do not log stack trace for known errors https://github.com/dotnet/aspnetcore/issues/29801
+                if (ex is not IOException && ex.InnerException is not AddressInUseException)
+                {
+                    Trace.LogCritical(0, ex, "Unable to start Kestrel.");
+                }
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/KestrelServerImpl.cs:215]: @davidfowl Was your suggestion to no longer log any startup exceptions? It feels wrong to log everything except AddressInUseExceptions.

Comment (davidfowl) [reply src/Servers/Kestrel/Core/src/Internal/KestrelServerImpl.cs:215]: Yep, don't log anything here. Just rethrow the exception.

**Review by halter73 — APPROVED** [review]

**Comment by halter73** [comment]
Thanks @BerserkerDotNet!

**Comment by maxcherednik** [comment]
@davidfowl I guess we addressed one of the issues mentioned here: #29801

To be specific - `log and throw` issue.

There was another issue, which is bigger than the one addressed in this pr. 

Can we please consider reopening the issue?

**Comment by davidfowl** [comment]
@maxcherednik , open a new issue please with details.

**Pull request merged**